pyomo/solver/ipopt.py

```diff
@@ -0,0 +1,517 @@
+                    remove_named_expressions=True,
+                )
+
+        # Capture/record end-time / wall-time
+        end_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        results.timing_info.start_timestamp = start_timestamp
+        results.timing_info.wall_time = (
+            end_timestamp - start_timestamp
+        ).total_seconds()
+        if config.report_timing:
+            results.report_timing()
+        return results
+
+    def _parse_ipopt_output(self, stream: io.StringIO):
+        """
+        Parse an IPOPT output file and return:
+
+        * number of iterations
+        * time in IPOPT
+
+        """
+
+        iters = None
+        nofunc_time = None
+        func_time = None
+        # parse the output stream to get the iteration count and solver time
+        for line in stream.getvalue().splitlines():
+            if line.startswith("Number of Iterations....:"):
+                tokens = line.split()
+                iters = int(tokens[3])
+            elif line.startswith(
+                "Total CPU secs in IPOPT (w/o function evaluations)   ="
+            ):
+                tokens = line.split()
+                nofunc_time = float(tokens[9])
+            elif line.startswith(
+                "Total CPU secs in NLP function evaluations           ="
+            ):
+                tokens = line.split()
+                func_time = float(tokens[8])
+
+        return iters, nofunc_time, func_time
+
+    def _parse_solution(
+        self, instream: io.TextIOBase, nl_info: NLWriterInfo, result: ipoptResults
+    ):
+        suffixes_to_read = ['dual', 'ipopt_zL_out', 'ipopt_zU_out']
+        res, sol_data = parse_sol_file(
+            sol_file=instream,
+            nl_info=nl_info,
+            suffixes_to_read=suffixes_to_read,
+            result=result,
+        )
+
+        if res.solution_status == SolutionStatus.noSolution:
+            res.solution_loader = SolutionLoader(None, None, None, None)
+        else:
+            rc = dict()
+            for v in nl_info.variables:
+                v_id = id(v)
+                rc[v_id] = (v, 0)
+                if v_id in sol_data.var_suffixes['ipopt_zL_out']:
+                    zl = sol_data.var_suffixes['ipopt_zL_out'][v_id][1]
+                    if abs(zl) > abs(rc[v_id][1]):
+                        rc[v_id] = (v, zl)
+                if v_id in sol_data.var_suffixes['ipopt_zU_out']:
+                    zu = sol_data.var_suffixes['ipopt_zU_out'][v_id][1]
+                    if abs(zu) > abs(rc[v_id][1]):
+                        rc[v_id] = (v, zu)
+
+            if len(nl_info.eliminated_vars) > 0:
+                sub_map = {k: v[1] for k, v in sol_data.primals.items()}
+                for v, v_expr in nl_info.eliminated_vars:
+                    val = evaluate_ampl_repn(v_expr, sub_map)
+                    v_id = id(v)
+                    sub_map[v_id] = val
+                    sol_data.primals[v_id] = (v, val)
```

Comment:
Maybe should be done by the `SolutionLoader` holding onto a copy of the `nl_info` and then do this as part of `get_primals`.

Reply:
I think we should just do this in `parse_sol_file`

Reply:
It should definitely not be here - that is agreed.